crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -0,0 +1,985 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::seq::IteratorRandom;
+use restate_core::metadata_store::{MetadataStoreClient, Precondition, ReadError, WriteError};
+use restate_core::network::NetworkSender;
+use restate_core::{ShutdownError, SyncError, TaskCenter, TaskKind};
+use restate_types::cluster::cluster_state::{ClusterState, NodeState, RunMode};
+use restate_types::cluster_controller::{
+    ReplicationStrategy, SchedulingPlan, SchedulingPlanBuilder,
+};
+use restate_types::identifiers::PartitionId;
+use restate_types::metadata_store::keys::SCHEDULING_PLAN_KEY;
+use restate_types::net::cluster_controller::Action;
+use restate_types::net::partition_processor_manager::{
+    ControlProcessor, ControlProcessors, ProcessorCommand,
+};
+use restate_types::{GenerationalNodeId, PlainNodeId, Version, Versioned};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+use tracing::{debug, trace};
+
+#[derive(Debug, thiserror::Error)]
+#[error("failed reading scheduling plan from metadata store: {0}")]
+pub struct BuildError(#[from] ReadError);
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("failed writing to metadata store: {0}")]
+    MetadataStoreWrite(#[from] WriteError),
+    #[error("failed reading from metadata store: {0}")]
+    MetadataStoreRead(#[from] ReadError),
+    #[error("failed syncing metadata: {0}")]
+    Metadata(#[from] SyncError),
+    #[error("system is shutting down")]
+    Shutdown(#[from] ShutdownError),
+}
+
+static EMPTY_NODE_SET: BTreeSet<PlainNodeId> = BTreeSet::new();
```

Comment:
Tip: for true static performance:
```
static EMPTY_NODE_SET: BTreeSet<PlainNodeId> = const { BTreeSet::new() };
```

Reply:
Interesting. I didn't know this difference. Thanks for making me aware of it :-)